sw/host/opentitanlib/src/io/emu.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transport::Result;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// Error related to the Simulator interface. These error messages will be printed in the context of
+/// a TransportError::EmuError, that is "EMU error: {}".  So including the words "error" or
+/// "Emu" in texts below will probably be redundant.
+#[derive(Error, Debug, Serialize, Deserialize)]
+pub enum EmuError {
+    #[error("Invalid argument name {0}")]
+    InvalidArgumetName(String),
+    #[error("Argument {0} has invalid value {1}")]
+    InvalidArgumentValue(String, String),
+    #[error("Start failed with cause {0}")]
+    StartFailureCause(String),
+    #[error("Stop failed with cause {0}")]
+    StopFailureCause(String),
+    #[error("Runtime error {0}")]
+    RuntimeError(String),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct EmuArgs {
```

Comment:
All 3 case are necessary to create HE instance.
1-2) list of app to execute
3) debug flag and some switches with don't have a value but have name.

Reply:
Some complicated container (trees / hash-map) will be problematic during serializations. Also container need to be ordered, some emulator sub-process could have argument parsing sensitive to argument order.

Reply:
IMHO: Blob like data should be transferred separately.

Reply:
Emulators may or may not involve a sub-process, some emulators could even utilize more than one sub-process.  For that reason, I think it is unwise to design EmuArgs be able to represent the arguments for exactly one shell command.

The document below mentions keys such as "flash", "app", "rom" to be used for various sections of the emulated storage.  There is not need to convey any ordering between those, and when we need to support multiple apps, then the `StringList` I propose would suffice.

I have not checked the JSON serialization format of `BTreeMap<K, V>`, but I had hoped that it would be identical to `Vec<(K, V)>` (or some variation with named fields `key` and `value`.)

https://docs.google.com/document/d/149qVPbBMWC5EtRM9oJHm-6mS_qAe3-L5Yk-RyZwL3Uc/edit?resourcekey=0-GkxvWco1yHiWs5W13KB6XA

Reply:
Thinking about it. Not all emulators need to pass the image in the same way, it could be that a future Verilator implementation wants the image (kernel and apps) in a single `EmuArg::Blob`, whereas the host emulator takes filesystem paths of kernel in `EmuArg::String` and paths to apps in a separate `EmuArg::StringList`.  If done that way, we could support remote verilator instances, without need for shared filesystem or separate file upload, and the host emulation could take advantage of the shared filesystem to directly execute the given binary.

Reply:
StringList is already implemented in last commit.
About tree/map i will check and see if it works.

There is my proposition for sending blob:
```
emulator send build/ti50.bin build/flash.img build/gpio_init.json
...
emulator start exec=ti50.bin flash=flash.img gpio=gpio_init.json
emulator stop
...
// update only one file
emulator send build/ti50_v2.1.bin
emulator start exec=ti50_v2.1.bin flash=flash.img build/gpio_init.json
...
emulator stop
emulator reset flash.img
emulator start exec=ti50_v2.1.bin flash=flash.img build/gpio_init.json
...
```

Reply:
I agree with @jesultra here that storing the data in some sort of key-value data structure seems to provide more flexibility to the architecture of the underlying "emulator", whether it be the Host Emulator, or a future Verilator subprocess. I would think that even if the emulator sub-process could have argument parsing sensitive to argument order, that the session proxy would know what that order was for the particular emulator backend, and could order things accordingly. Moreover, if that is not possible, @cfrantz mentioned the [indexmap](https://docs.rs/indexmap/latest/indexmap/) crate that could be a viable compromise.

Reply:
@timothytrippel 
I added `args: &HashMap<String, EmuValue>`  to start signature.